src/syntaxtree.c

```diff
@@ -0,0 +1,850 @@
+        }
+    }
+
+    AssPRec(result, RNamName("list"), list);
+
+    return result;
+}
+
+static Obj SyntaxTreeRangeExpr(Obj result, Expr expr)
+{
+    Obj first;
+    Obj second;
+    Obj last;
+
+    if (SIZE_EXPR(expr) == 2 * sizeof(Expr)) {
+        first = SyntaxTreeCompiler(READ_EXPR(expr,0));
+        last = SyntaxTreeCompiler(READ_EXPR(expr,1));
+
+        AssPRec(result, RNamName("first"), first);
+        AssPRec(result, RNamName("last"), last);
+    }
+    else {
+        first = SyntaxTreeCompiler(READ_EXPR(expr,0));
+        second = SyntaxTreeCompiler(READ_EXPR(expr,1));
+        last = SyntaxTreeCompiler(READ_EXPR(expr,2));
+
+        AssPRec(result, RNamName("first"), first);
+        AssPRec(result, RNamName("second"), second);
+        AssPRec(result, RNamName("last"), last);
+    }
+
+    return result;
+}
+
+// TODO: Can we use EvalStringExpr? Indeed can we use EvalXexpr
+//       for more/all "basic" datatypes?
+static Obj SyntaxTreeStringExpr(Obj result, Expr expr)
+{
+    Obj string;
+    UInt len;
+
+    len = READ_EXPR(expr, 0);
+    string = NEW_STRING(len);
+    memcpy(ADDR_OBJ(string), CONST_ADDR_EXPR(expr), SIZEBAG_STRINGLEN(len));
+
+    AssPRec(result, RNamName("string"), string);
+
+    return result;
+}
+
+static Obj SyntaxTreeRecExpr(Obj result, Expr expr)
+{
+    Obj  key;
+    Obj  val;
+    Obj  list;
+    Obj  subrec;
+    Expr tmp;
+    Int  i, len;
+
+    len = SIZE_EXPR(expr) / (2 * sizeof(Expr));
+    list = NEW_PLIST(T_PLIST, len);
+    SET_LEN_PLIST(list, len);
+
+    for (i = 1; i <= len; i++) {
+        tmp = READ_EXPR(expr,2 * i - 2);
+        GAP_ASSERT(tmp != 0);
+
+        subrec = NEW_PREC(2);
+        SET_ELM_PLIST(list, i, subrec);
+        CHANGED_BAG(list);
+
+        if (IS_INTEXPR(tmp)) {
+            key = NAME_RNAM((UInt)INT_INTEXPR(tmp));
+        }
+        else {
+            key = SyntaxTreeCompiler(tmp);
+        }
+        AssPRec(subrec, RNamName("key"), key);
+
+        tmp = READ_EXPR(expr,2 * i - 1);
+        val = SyntaxTreeCompiler(tmp);
+        AssPRec(subrec, RNamName("value"), val);
+    }
+    AssPRec(result, RNamName("keyvalue"), list);
+
+    return result;
+}
+
+/* TODO: Slightly ugly hack */
+extern Obj EAGER_FLOAT_LITERAL_CACHE;
+static Obj SyntaxTreeFloatEager(Obj result, Expr expr)
+{
+    UInt ix = READ_EXPR(expr, 0);
+    AssPRec(result, RNamName("value"), ELM_LIST(EAGER_FLOAT_LITERAL_CACHE, ix));
+    return result;
+}
+
+static Obj SyntaxTreeFloatLazy(Obj result, Expr expr)
+{
+    UInt len;
+    Obj string;
+
+    len = READ_EXPR(expr, 0);
+    string = NEW_STRING(len);
+    memcpy(CHARS_STRING(string),
+           (const char *)CONST_ADDR_EXPR(expr) + 2 * sizeof(UInt), len);
+    AssPRec(result, RNamName("value"), string);
+    return result;
+}
+
+static Obj SyntaxTreeRefLVar(Obj result, Expr expr)
+{
+    LVar lvar;
+
+    if (IS_REFLVAR(expr)) {
+        lvar = LVAR_REFLVAR(expr);
+    } else {
+        lvar = (LVar)(READ_EXPR(expr,0));
+    }
```

Comment:
OK, I see you have folded `SyntaxTreeLVar` in here now. But I am still confused as to why you handle `T_REFLVAR` and `T_ISB_LVAR` via `SyntaxTreeRefLVar`, but then also have
```C
    COMPILER_(T_ASS_LVAR,
               ARG("lvar", SyntaxTreeIntObjInt), ARG_("rhs")),
     COMPILER_(T_UNB_LVAR, ARG("lvar", SyntaxTreeIntObjInt)),
```
I would expect all three to be handled uniformly, and the above code put into a `SyntaxTreeArgcompLVar` helper (which would probably be identical to `SyntaxTreeIntObjInt`)

Reply:
Such code would also break the executor, as in `vars.c`, there is no such check. This seems to be a "bug" in the compiler code (which was written later)

Reply:
@fingolfin figured out that this code used to handle references to `LVar`s before they were changed in 8d2ca938c2f4102208b72e21a87d723f9d8446e0